fix: Enable access to block settings within UBE

dcalhoun opened the pull request:
## Related PRs

* https://github.com/wordpress-mobile/gutenberg-mobile/pull/5508

<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Repair access to the block settings sidebar within the unsupported block editor
(UBE). Fixes #48298. 

Repair hiding of unrelated editor UI in the UBE. Fixes #47966. 

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
Without access to the block settings, a user cannot modify important settings
for a given block. 

Also, displaying UI elements unrelated to the sole block is
confusing and unnecessary. 

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

### Block Settings
Previously, the UBE relied upon the presence of the "Show more settings" button,
but that was removed in #46709. This updates the UBE to rely upon the sidebar
toggle button that is present in the post editor header. Because the sidebar toggle 
button can also reveal the post settings, this only displays the sidebar toggle 
button whenever a block is selected.

### Post Title Visibility

The override styles used by the UBE were no longer successfully applied once https://github.com/WordPress/gutenberg/pull/46212 
wrapped the editor within an `iframe`. It is not possible to style `iframe` 
children elements from the parent `document` context. These changes no 
copy the override styles to reside within the `iframe` document so they are 
applied to the children elements. 

### Unrelated Changes

This also updates existing Android-specific scripts to only run when Android
devices are detected by parsing the user agent string. Previously, the entire
`editor-behavior-overrides` script was only loaded by the Android platform. 
Now, both platforms load the script, and portions of it only run for Android. 

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

> **Note**
> There is one known issue. On iOS, tapping towards the bottom of the screen — between the sole block and the keyboard toolbar — clears the block selection, but the sidebar toggle button remains erroneously displayed. This does not occur for Android. This can be seen at the end of the screencast below. 

Verify the steps outlined in #48298, #34668, and #47966 pass.

1. Using the WordPress mobile app, open a post containing unsupported blocks,
   e.g. Archives, Table.
1. Select an unsupported block.
1. Tap the <kbd>?</kbd> button.
1. Tap "Edit using web editor."
1. Verify the sidebar toggle button is not present in the top-right corner.
1. Select the unsupported block.
1. Tap the sidebar toggle button in the top-right corner.
1. Verify the block settings are displayed.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for

 PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->
n/a

## Screenshots or screencast <!-- if applicable -->

https://user-images.githubusercontent.com/438664/221731846-229b3e08-321d-417f-aa4e-68b1338ec86e.mov

dcalhoun committed:
fix: Enable access to block settings within UBE

The "Show more settings" menu item is no longer included in the block
toolbar after #46709 merged. Rather than relying upon that menu item,
this conditionally displays the sidebar toggle button whenever a block
is selected.

dcalhoun committed:
fix: Disable white space stripping that breaks nested CSS selectors

CSS selectors rely upon a single white space between selectors to
represent an ancestor relationship. Globally removing white space in the
stylesheet breaks this functionality, as it transforms the selector to
target a single element with all the selectors.

The white space stripping should likely be replaced with a proper CSS
minification long term.

dcalhoun committed:
fix: Hide block actions unrelated to editing a single block

Hide the entire "block settings" drop-down menu now that we no longer
rely upon it to access the "Show more settings" menu option that was
removed entirely.

dcalhoun committed:
refactor: Relocate script toggling block settings visibility

This relates more to editor behavior than the post content.

dcalhoun commented:
Hey @geriux. 👋🏻 There is one known issue referenced in the PR description, but I wanted to seek your feedback on the approach here. Also, if you have any ideas on known issue let me know. 😬😄 Thanks!

dcalhoun commented:
For the known issue, it appears [the logic](https://github.com/WordPress/gutenberg/blob/f64379314802d2d4b64077ca6b980e9df824b762/packages/block-editor/src/components/block-selection-clearer/index.js) clearing block selection does dispatch `CLEAR_SELECTED_BLOCK`, but for whatever reason that is not received by the store extending logic in this PR. Clicking elsewhere _does_ cause the dispatched event to arrive to the store extending logic. 

Additionally, I attempted to leverage the [`clearBlockSelection` settings](https://github.com/WordPress/gutenberg/blob/4b4b4ca96f201a06cf7fffb9921041957f070fac/packages/block-editor/src/store/defaults.js#L30) to prevent removing selection from the sole block. However, the [setting value](https://github.com/WordPress/gutenberg/blob/f64379314802d2d4b64077ca6b980e9df824b762/packages/block-editor/src/components/block-selection-clearer/index.js#L23) is stale/incorrect _only_ when clicking in this space between the title and block toolbar. 

It is like this space is a "twilight zone." 😄 Not sure what is going on.

geriux commented:
> There is one known issue. On iOS, tapping in-between the block toolbar and the post title clears the block selection, but the sidebar toggle button remains erroneously displayed. This does not occur for Android. This can be seen at the end of the screencast below.

One thing that noticed from this comment was the mention of the post title and then I realized from testing and seeing the demo video that the Post title and the (+) Button are there.

I guess something changed in Gutenberg Web and how it loads the editor but it looks like other things we have for the UBE are not working correctly, like the CSS rules to hide the Post title, (+) button, etc.

Maybe if we check why that's not working it could solve the issue you mentioned about the post title clearing the selection, what do you think?

I looked for an older screenshot of the UBE and this is what it looked like:

<kbd><img width="156" alt="Screenshot 2023-02-27 at 13 10 17" src="https://user-images.githubusercontent.com/4885740/221560494-94b95c63-9c4f-406b-9c77-eaec4f98a2b1.png"></kbd>

As you can see only the toolbar and the block are visible.

dcalhoun commented:
> One thing that noticed from this comment was the mention of the post title and then I realized from testing and seeing the demo video that the Post title and the (+) Button are there.

Yes, I noticed the same and [investigated it a bit](https://github.com/WordPress/gutenberg/issues/47966#issuecomment-1444264356). 

> Maybe if we check why that's not working it could solve the issue you mentioned about the post title clearing the selection, what do you think?

I had considered them separate issues, but you are right that solving https://github.com/WordPress/gutenberg/issues/47966 might have an impact on the clear selected block issue. I'll look into it more. 👀

geriux commented:
> Yes, I noticed the same and [investigated it a bit](https://github.com/WordPress/gutenberg/issues/47966#issuecomment-1444264356).
> >

Ohh I missed that issue, thanks for sharing the link 😃 

>I had considered them separate issues, but you are right that solving https://github.com/WordPress/gutenberg/issues/47966 might have an impact on the clear selected block issue. I'll look into it more. 👀

I wonder if we could use the iFrame's address to render the content in the UBE directly 🤔 I haven't tested this but maybe it could be a possible fix if trying to inject content within the iframe becomes difficult.

geriux commented:
> I wonder if we could use the iFrame's address to render the content in the UBE directly 🤔 I haven't tested this but maybe it could be a possible fix if trying to inject content within the iframe becomes difficult.

Nevermind, it is setting the editor's content using `srcdoc` instead of an URL 😅

dcalhoun committed:
fix: Apply styles and script to editor canvas iframe

The editor canvas now relies upon an iframe. It is not possible to style
elements within an iframe from the parent context. This copies the
styles from the parent conext to the iframe.

Additionally, the logic selecting the first block also failed due to the
block not existing when it was invoked. This relocates that logic until
after the iframe is ready.

dcalhoun committed:
fix: Expand conditional checks for partial DOM trees

On Android, there were times where the iframe was present, but the
nested window was not yet ready.

dcalhoun committed:
refactor: Rename for brevity

dcalhoun committed:
fix: Avoid React removing appended iframe styles

Append the styles to the `document` element, as React will remove the
mutation to the `head` element.

dcalhoun commented on packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/GutenbergWebViewActivity.java:
```diff
@@ -327,16 +327,16 @@ private void injectCssScript() {
         mWebView.evaluateJavascript(injectCssScript, message -> {
             if (message != null) {
                 String editorStyle = getFileContentFromAssets("gutenberg-web-single-block/editor-style-overrides.css");
-                editorStyle = removeWhiteSpace(removeNewLines(editorStyle));
```
This rudimentary stripping of whitespace inadvertently modifies the CSS selectors, causing them to be interpreted as completely different selectors. E.g. `.ancestor .descendent` is different from `.sibling-1.sibling-2`. Ideally, a proper CSS minification is utilized long term.

dcalhoun commented on packages/react-native-bridge/common/gutenberg-web-single-block/content-functions.js:
```diff
@@ -25,19 +25,12 @@ window.getHTMLPostContent = () => {
 		.setupEditor( post, { content: blockHTML } );
 
-	// Select the first block.
-	const clientId = blockEditorSelect.getBlocks()[ 0 ].clientId;
```
This often threw an error where the first block was `undefined`. This logic was relocated to `editor-behavior-overrides` to await the `iframe` render.

dcalhoun commented on packages/react-native-bridge/common/gutenberg-web-single-block/editor-behavior-overrides.js:
```diff
@@ -1,59 +1,167 @@
+
+window.wp.data.use( toggleBlockSelectedStyles );
+
+// The editor-canvas iframe relies upon `srcdoc`, which does not trigger a
+// `load` event. Thus, we must poll for the iframe to be ready.
+let overrideAttempts = 0;
+const overrideInterval = setInterval( () => {
```
I explored a few different approaches for awaiting the `iframe` render. Unfortunately, none of them succeeded.

- `iframe.onload` does not work when the `iframe` uses `srcdoc` instead of `src`. 
- Awaiting a "ready Gutenberg" via [subscribing to the store](https://github.com/WordPress/gutenberg/blob/a7eb9a220f7105d6a194b0e0228fbbe2484ae727/packages/react-native-bridge/common/gutenberg-web-single-block/content-functions.js#L60-L68) runs before the `iframe` renders.

dcalhoun commented on packages/react-native-bridge/common/gutenberg-web-single-block/editor-behavior-overrides.js:
```diff
@@ -1,59 +1,167 @@
+		// elements within an iframe cannot be styled from the parent context.
+		const overrideStylesClone = overrideStyles.cloneNode( true );
+		overrideStylesClone.id = 'editor-styles-overrides-2';
+		// Append to document rather than the head, as React will remove this
+		// mutation.
+		canvasIframe.contentDocument.documentElement.appendChild(
```
The `iframe` implementation "manages" the `head` element. So, we must avoid mutating that element, otherwise React may remove any changes we make.

dcalhoun committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into fix/enable-access-to-block-settings-within-ube

dcalhoun commented:
👋🏻 @geriux. I updated this PR to address the missing styles hiding the unrelated editor UI. Unfortunately, doing so did not resolve the issue where tapping certain areas of the do not properly trigger a `CLEAR_SELECTED_BLOCK` action and toggle the sidebar button. 

That said, while I am not too proud of the implementation presented in this PR, I am wondering it makes sense to land this in a beta fix for the time being. It improves the UX a good bit IMO and we are approach the app release. If desired, we could iterate further on the implementation at a future time.  

For future iterations, there are a lot of oddities that occur due to the new `iframe`. It is a bit of a "dance" to get the UBE functioning as we expect. It may be worthwhile to push again for a first-party mechanism for rendering a single block with a streamlined interface to support it that doesn't involve overriding existing styles and scripts.

geriux commented:
> That said, while I am not too proud of the implementation presented in this PR, I am wondering it makes sense to land this in a beta fix for the time being. It improves the UX a good bit IMO and we are approach the app release. If desired, we could iterate further on the implementation at a future time.

I agree we should land this for now and see what we do afterward, I'll focus on reviewing and testing today, thanks!

>For future iterations, there are a lot of oddities that occur due to the new iframe. It is a bit of a "dance" to get the UBE functioning as we expect. It may be worthwhile to push again for a first-party mechanism for rendering a single block with a streamlined interface to support it that doesn't involve overriding existing styles and scripts.

I was thinking maybe we could add a flag so it doesn't load the iFrame for the UBE, I saw this PR so maybe we [could do the same](https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-post/src/components/visual-editor/index.js#L342-L350).

dcalhoun commented:
> I agree we should land this for now and see what we do afterward, I'll focus on reviewing and testing today, thanks!

Great — thanks! 

> I was thinking maybe we could add a flag so it doesn't load the iFrame for the UBE, I saw this PR so maybe we [could do the same](https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-post/src/components/visual-editor/index.js#L342-L350).

Possibly. I did see the flag as well. However, It seems the [`iframe ` initiative](https://github.com/WordPress/gutenberg/issues/33346) is being done for very specific reasons, some of which are outlined in more detail in https://github.com/WordPress/gutenberg/issues/20797. While not all the reasons may be relevant for the UBE, some of them may be relevant. E.g. avoiding leaked admin styles overriding theme styles in the editor. So, I am unsure if disabling the `iframe` is a good long-term solution or if the flag will even be available long term.

dcalhoun committed:
docs: Add change log entry

geriux reviewed (approved):
LGTM! 🚀  Nice work and thanks for working on a fix so quickly 👏 

The new code and organization changes make sense to me, I also tested this on both iOS and Android, and I shared what I tested over this [comment](https://github.com/wordpress-mobile/gutenberg-mobile/pull/5508#issuecomment-1448584811) in Gutenberg mobile.
